packages/ckeditor5-html-support/src/htmlcomment.js

```diff
@@ -88,6 +88,35 @@ export default class HtmlComment extends Plugin {
 
 			return true;
 		} );
+
+		// Delete all comment markers from the document before setting new data.
+		editor.data.on( 'set', () => {
+			for ( const commentMarker of editor.model.markers.getMarkersGroup( '$comment' ) ) {
+				this.removeHtmlComment( commentMarker.name );
+			}
+		}, { priority: 'high' } );
+
+		// Delete all comment markers that are within a removed range.
+		// Delete all comment markers at the limit element boundaries if the whole content of the limit element is removed.
+		editor.model.on( 'deleteContent', ( evt, [ selection ] ) => {
+			for ( const range of selection.getRanges() ) {
+				const limitElement = editor.model.schema.getLimitElement( range );
+				const firstPosition = editor.model.createPositionAt( limitElement, 0 );
+				const lastPosition = editor.model.createPositionAt( limitElement, 'end' );
+
+				let affectedCommentIDs;
+
+				if ( firstPosition.isTouching( range.start ) && lastPosition.isTouching( range.end ) ) {
+					affectedCommentIDs = this.getHtmlCommentsInRange( editor.model.createRange( firstPosition, lastPosition ) );
+				} else {
+					affectedCommentIDs = this.getHtmlCommentsInRange( range, { skipBoundaries: true } );
+				}
```

Comment:
Maybe these few lines could be simplified a bit?

```suggestion
				// Delete comment markers at range boundaries if limit element's boundaries touch the start and end of the removed range.
				const skipBoundaries = !firstPosition.isTouching( range.start ) || !lastPosition.isTouching( range.end );
				const affectedCommentIDs = this.getHtmlCommentsInRange( range, { skipBoundaries } );
```

Reply:
Note that the range differs in both cases.

Reply:
If both positions touch the boundary of the limit element then these two positions are used to create a range instead of the original ones.

Reply:
You are right. I assumed that `position.isTouching()` behavies like `position.isEqual()`, but there is a slight difference between them.